File: 307. Range Sum Query - Mutable/307. Range Sum Query - Mutable.js
class MyTreeNode{
  constructor(nums){
    const l = nums.length;
    const m = l >> 1;
    this.val = nums[m];
    this.changed = false;
    this.sum = this.val;
    if(m === 0){
      this.leftCount = 0;
      this.left = null;
    }else{
      this.leftCount = m;
      this.left = new MyTreeNode(nums.slice(0, m));
      this.sum += this.left.sum;
    }
    if(m === l-1){
      this.rightCount = 0;
      this.right = null;
    }else{
      this.rightCount = l-m-1;
      this.right = new MyTreeNode(nums.slice(m+1));
      this.sum += this.right.sum;
    }
  }
}

class MyTree{
/**
 * @param {number[]} nums
 */
  constructor(nums){
    if(nums.length){
      this.root = new MyTreeNode(nums);
    }else{
      this.root = null;
    }
    
    this.changed = false;
  }
  
/** 
 * @param {number} i 
 * @param {number} val
 * @return {void}
 */
  update(i, val){
    this.changed = true;
    let p = this.root;
    for(;;){
      console.log(i, p.val);
      if(p === null){
        throw new Error('can\'t find this node');
      }
      p.changed = true;
      if(p.leftCount === i){
        p.val = val;
        break;
      }else if(i < p.leftCount){
        p = p.left;
      }else{
        i -= p.leftCount + 1;
        p = p.right;
      }
    }
  }
  maintain(){
    if(this.changed){
      this.change = false;
      function getSum(node){
        if(node === null){
          return 0;
        }
        if(node.changed){
          node.sum = getSum(node.left) + node.val + getSum(node.right);
          node.change = false;
          return node.sum;
        }else{
          return node.sum;
        }
      }
      getSum(this.root);
    }
  }

/** 
 * @param {number} i 
 * @param {number} j
 * @return {number}
 */
  sumRange(i, j){
    this.maintain();
    function getSum(node, i, j){
      if(i === j && i === node.leftCount){
        return node.val;
      }
      if(i === 0 && j === (node.leftCount + node.rightCount)){
         return node.sum;
      }
      if(node.leftCount < i){
        // mid to the left of [i, j]
        return getSum(node.right, i - node.leftCount - 1, j - node.leftCount - 1);
      }else if(node.leftCount === i){
        // mid === i
        return node.val + getSum(node.right, 0, j - node.leftCount - 1);
      }else if(i < node.leftCount && node.leftCount < j){
        // mid in [i, j]
        return getSum(node.left, i, node.leftCount - 1) + node.val + getSum(node.right, 0, j - node.leftCount - 1);
      }else if(j === node.leftCount){
        // mid === j
        return getSum(node.left, i, node.leftCount - 1) + node.val;
      }else if(j < node.leftCount){
        // mid to the right of [i, j]
        return getSum(node.left, i, j); 
      }
      
      throw new Error('out of cases');
    }
    return getSum(this.root, i, j);
  }
}

var NumArray = MyTree;

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = Object.create(NumArray).createNew(nums)
 * obj.update(i,val)
 * var param_2 = obj.sumRange(i,j)
 */

File: 307. Range Sum Query - Mutable/307. Range Sum Query - Mutable.rs
use std::ops::Add;
type Pred = fn(i32, i32) -> i32;

struct SegTree {
    len: usize,
    data: Vec<i32>,
    func: Pred,
}

impl SegTree {
    fn new(data: &[i32], func: Pred) -> Self {
        let len = data.len();
        let mut st = Self {
            len,
            data: vec![0; 4 * len],
            func,
        };
        st._build(data, 0, 0, len);
        st
    }
    fn _build(&mut self, data: &[i32], node: usize, l: usize, r: usize) {
        if l + 1 == r {
            self.data[node] = data[l];
        } else {
            let mid = (l + r) / 2;
            let ln = node * 2 + 1;
            let rn = node * 2 + 2;
            self._build(data, ln, l, mid);
            self._build(data, rn, mid, r);
            self.data[node] = (self.func)(self.data[ln], self.data[rn]);
        }
    }
    fn query(&self, l: usize, r: usize) -> i32 {
        self._q(0, 0, self.len, l, r).unwrap()
    }
    fn _q(&self, node: usize, nl: usize, nr: usize, l: usize, r: usize) -> Option<i32> {
        // nl, nr:  data[node] covers range [nl..mr]
        if r <= nl || nr <= l {
            return None;
        }
        if l <= nl && nr <= r {
            return Some(self.data[node]);
        }
        let nm = (nl + nr) / 2;
        let cl = self._q(node * 2 + 1, nl, nm, l, r);
        let cr = self._q(node * 2 + 2, nm, nr, l, r);
        if let Some((l, r)) = cl.zip(cr) {
            Some((self.func)(l, r))
        } else {
            cl.or(cr)
        }
    }

    fn update(&mut self, value: i32, i: usize) {
        self._u(0, 0, self.len, value, i);
    }
    fn _u(&mut self, node: usize, nl: usize, nr: usize, value: i32, i: usize) {
        if i < nl || nr <= i {
            return;
        }
        if nl == i && nl + 1 == nr {
            self.data[node] = value;
            return;
        }
        let nm = (nl + nr) / 2;
        let cl = node * 2 + 1;
        let cr = node * 2 + 2;
        self._u(cl, nl, nm, value, i);
        self._u(cr, nm, nr, value, i);
        self.data[node] = (self.func)(self.data[cl], self.data[cr]);
    }
}


struct NumArray {
    st: SegTree,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumArray {
    fn new(nums: Vec<i32>) -> Self {
        Self {
            st: SegTree::new(&nums, Add::add),
        }
    }
    
    fn update(&mut self, index: i32, val: i32) {
        self.st.update(val, index as usize);
    }
    
    fn sum_range(&self, left: i32, right: i32) -> i32 {
        self.st.query(left as usize, right as usize + 1)
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * let obj = NumArray::new(nums);
 * obj.update(index, val);
 * let ret_2: i32 = obj.sum_range(left, right);
 */
